### ISSUE TITLE
2D tf.nn.convolution() output is inconsistent between inputs 1x1 and 2x2 with kernel=3x3, strides=2, padding='SAME'.

### ISSUE THREAD

**Issue by ozabluda** [tensorflow/tensorflow#17654]
Output of Keras `Conv2D()`, which is a passthrough to `tf.nn.convolution()`, is inconsistent between inputs 1x1 and 2x2 with kernel=3x3, strides=2, padding='SAME'.

The following code with 2x2 input produces wrong output `8.`, which is inconsistent with correct output `4.` if the input is 1x1. 
```
import numpy as np
from keras.layers import Conv2D
from keras.models import Sequential

w = np.arange(3*3*1*1).reshape(3, 3, 1, 1)

model = Sequential()
model.add(Conv2D(1, kernel_size=3, strides=2, input_shape=(2, 2, 1),
                 padding='SAME', use_bias=False))
model.set_weights([w])
print(model.predict(np.ones((1, 2, 2, 1))))
```

**Comment by ozabluda** [comment]
For more information, see https://github.com/keras-team/keras/pull/9473#issuecomment-372166860, which narrowed it down to "odd kernel / even input", and also found inconsistency with Theano and CNTK.

**Comment by ppwwyyxx** [comment]
Following the [documentation](https://www.tensorflow.org/versions/r1.0/api_guides/python/nn#Convolution), the correct output is 0+1+3+4=8. So it's working as expected.

**Comment by ozabluda** [comment]
Which documentation is it? As far as I can tell, the difference is entirely due to difference in padding. In the case of 1x1 input, the padded input looks like so:
```
0 0 0
0 1 0
0 0 0
```
but in case of 3x3 input, the padded input looks like so:
```
1 1 0 0
1 1 0 0
0 0 0 0
0 0 0 0
```
while I expected it to look like so (and this is indeed how it is with `strides=1`):
```
0 0 0 0 
0 1 1 0
0 1 1 0
0 0 0 0
```
An easy way to see it is by adding `w = np.power(2, w)`

The closest I see documentation-wise is
>padded_input is obtained by zero padding the input using an effective spatial filter shape of (spatial_filter_shape-1) * dilation_rate + 1 and output striding strides as described in the [comment here](https://www.tensorflow.org/api_docs/python/tf/nn/convolution).

https://www.tensorflow.org/api_docs/python/tf/nn/convolution

but that link is broken

**Comment by ozabluda** [comment]
For the reference, here is how a complete code producing the expected result `432` (to me the difference with above is unexpected):
```
import numpy as np
from keras.layers import Conv2D, ZeroPadding2D
from keras.models import Sequential


w = np.arange(3*3*1*1).reshape(3, 3, 1, 1)
w = np.power(2, w)

model = Sequential()
model.add(ZeroPadding2D(padding=(1, 1), input_shape=(2, 2, 1)))
model.add(Conv2D(1, kernel_size=3, strides=2, padding='VALID', use_bias=False))
model.set_weights([w])
print(model.predict(np.ones((1, 2, 2, 1))))
```

**Comment by ppwwyyxx** [comment]
I posted the link to documentation of "Convolution".
The documentation says the padded input is:
```
1 1 0
1 1 0
0 0 0
```

**Comment by ozabluda** [comment]
Oh, yeah, see it now. Thank you. Indeed it is:

https://www.tensorflow.org/api_guides/python/nn#Convolution
https://www.tensorflow.org/api_guides/python/nn#Notes_on_SAME_Convolution_Padding

```
in_height, in_width = 2,2
filter_height, filter_width = 3,3
strides=(None,2,2)
out_height = np.ceil(float(in_height) / float(strides[1]))
out_width  = np.ceil(float(in_width) / float(strides[2]))

#The total padding applied along the height and width is computed as:

if (in_height % strides[1] == 0):
  pad_along_height = max(filter_height - strides[1], 0)
else:
  pad_along_height = max(filter_height - (in_height % strides[1]), 0)
if (in_width % strides[2] == 0):
  pad_along_width = max(filter_width - strides[2], 0)
else:
  pad_along_width = max(filter_width - (in_width % strides[2]), 0)

print(pad_along_height, pad_along_width)
  
#Finally, the padding on the top, bottom, left and right are:

pad_top = pad_along_height // 2
pad_bottom = pad_along_height - pad_top
pad_left = pad_along_width // 2
pad_right = pad_along_width - pad_left

print(pad_top, pad_bottom, pad_left, pad_right)
```

output:
```
1 1
0 1 0 1
```
